remme/clients/basic.py

```diff
@@ -232,60 +213,276 @@ def _handle_response(self, msg_type, resp_proto, req):
-
-    def get_root_block(self):
-        resp = self._handle_response(Message.CLIENT_BLOCK_LIST_REQUEST,
-                                     ClientBlockListResponse,
-                                     ClientBlockListRequest(
-                                         paging=ClientPagingControls(limit=1)))
-
-        block = resp['blocks'][0]
-        header = BlockHeader()
-        try:
-            header_bytes = base64.b64decode(block['header'])
-            header.ParseFromString(header_bytes)
-        except (KeyError, TypeError, ValueError, DecodeError):
-            header = block.get('header', None)
-            LOGGER.error(
-                'The validator sent a resource with %s %s',
-                'a missing header' if header is None else 'an invalid header:',
-                header or '')
-            raise ClientException()
-
-        block['header'] = message_to_dict(header)
+    @classmethod
+    def _get_metadata(cls, response, head=None):
+        head = response.get('head_id', head)
+        metadata = {}
+        if head is not None:
+            metadata['head'] = head
+        return metadata
 
+    @staticmethod
+    def _wrap_response(data=None, metadata=None):
+        envelope = metadata or {}
+
+        if data is not None:
+            envelope['data'] = data
+
+        return envelope
+
+    @classmethod
+    def _wrap_paginated_response(cls, response, controls, data, head=None):
+        paging_response = response['paging']
+        if head is None:
+            head = response['head_id']
+
+        paging = {
+            'limit': controls.get('limit'),
+            'start': controls.get('start'),
+            'next': paging_response.get('next'),
+        }
+        return cls._wrap_response(
+            data=data,
+            metadata={
+                'head': head,
+                'paging': paging
+            })
+
+    def _head_to_root(self, block_id):
+        if block_id:
+            resp = self._handle_response(
+                Message.CLIENT_BLOCK_GET_BY_ID_REQUEST,
+                ClientBlockGetResponse,
+                ClientBlockGetByIdRequest(block_id=block_id)
+            )
+            block = expand_block(resp['block'])
+        else:
+            resp = self._handle_response(
+                Message.CLIENT_BLOCK_LIST_REQUEST,
+                ClientBlockListResponse,
+                ClientBlockListRequest(
+                    paging=ClientPagingControls(limit=1)
+                )
+            )
+            block = expand_block(resp['blocks'][0])
         return (
             block['header_signature'],
             block['header']['state_root_hash'],
         )
 
-    def submit_batches(self, data):
-        self._handle_response(Message.CLIENT_BATCH_SUBMIT_REQUEST,
-                              ClientBatchSubmitResponse,
-                              ClientBatchSubmitRequest(batches=data['batches']))
+    def list_state(self, address, start=None, limit=None, head=None,
```

Comment:
You handle too much logic in the basic class. Leave a `TODO` note here. I will be reviewing them soon.